### PR TITLE
Infer generics from typeargs of return types of closures.

### LINKED ISSUE #921: Rock can't pick up generic type info from the typeargs of the return value of closure
I know, quite the title.  

Here is what I mean:

```ooc

/* THIS WORKS */

foo: func <T> (c: Cell<T>) { T name println() }
foo(Cell new(42))

/* THIS WORKS TOO */

foo: func <T> (f: Func -> T) { T name println() }
foo(|| 42)  

/* THIS DOESN'T */

foo: func <T> (f: Func -> Cell<T>) { T name println() }
foo(|| Cell new(42))
```

The above fails with `error Missing info for type argument T. Have you forgotten to qualify foo<T> (f: Func () -> Cell), e.g. List<Int>?` (btw, this error message is pretty misleading to beginners since explicitly qualifying functions, rather than types, is illegal in ooc)

### PR DESCRIPTION
**Pull request by shamanas**
Closes #921

### CODE REVIEW

File: source/rock/middle/FunctionCall.ooc [hunk 1]
```diff
@@ -1255,6 +1259,32 @@ FunctionCall: class extends Expression {
                                     return null
                                 }
                             }
+                        } else if (retType) {
+                            if (implArg instanceOf?(FunctionDecl)) {
```
Comment (fasterthanlime) [inline source/rock/middle/FunctionCall.ooc]: `instanceOf? + as` is a form I avoid in rock code nowadays.

This one is preferred:

```ooc
match implArg {
  case fDecl: FunctionDecl =>
    argType := fDecl inferredReturnType
    // etc.
}
```

File: source/rock/middle/FunctionCall.ooc [hunk 2]
```diff
@@ -1255,6 +1259,32 @@ FunctionCall: class extends Expression {
...
+                                if (argType) {
+                                    // We don't return on the else case because we are not even sure we can qualify the type arg from the inferred return type
+
+                                    // Here is how we will proceed:
+                                    // We will find the index of our type arg name in the retType typeargs
+                                    // Then match it to the type at that index of the argType
+                                    returnTypeTypeargs := retType getTypeArgs()
+                                    for ((i, typeArg) in returnTypeTypeargs) {
+                                        if (typeArgName == typeArg getName()) {
+                                            //Yuup
```
Comment (fasterthanlime) [inline source/rock/middle/FunctionCall.ooc]: we can probably lose that line :)

File: source/rock/middle/FunctionCall.ooc (hunk 1 above)
Comment (shamanas) [inline source/rock/middle/FunctionCall.ooc]: I feel weird having a match with a single branch, but I guess I'll get used to it

**Comment by vendethiel** [comment]
Great work :o :+1:

**Pull request merged**